### PR TITLE
[Perf] Improve dirty AST rebuild speed 2.56x on many-core machines

### PR DESCRIPTION
**Pull request by davezarzycki**
On many-core machines, the files that take the longest to compile are build bottlenecks. With this change, we workaround clang scaling problems compiling large llvm::StringSwitch expressions and bring the `touch tools/swift/include/swift/AST/* ; time ninja swift` time down from 1m43.24s down to 40.26s.

I've also benchmarked rebuilding Swift.o after this change. The before and after seems to be deep in the noise.

### CODE REVIEW

**Review by jrose-apple — APPROVED** [review]
I wouldn't generally be in favor of removing StringSwitch, but given that we're already using xmacros here it doesn't seem to matter.

**Comment by davezarzycki** [comment]
For whatever it may be worth, the before and after difference for Builtins.cpp is crazy. With llvm::StringSwitch, it takes 80+ seconds to compile. After this change, it takes less than 4.

**Comment by gottesmm** [comment]
@davezarzycki This is a pattern that we use in a bunch of places. If it interests you, it may be worth it to just optimize StringSwitch in the general case. Why should it take so long to compile such a thing? Not sure.

**Comment by gottesmm** [comment]
Sorry let me rephrase. While in this case, it seems worth it to eliminate the use of StringSwitch and get the win, the StringSwitch issue sounds like a bigger problem. This is especially true since we use this same pattern in many other places in the compiler. It may be worth to investigate if StringSwitch or clang can be optimized.

**Comment by DougGregor** [comment]
Wow. I feel even worse now about introducing `StringSwitch`.

**Comment by jrose-apple** [comment]
I suspect it's not nearly such a problem when it's not an xmacro with dozens of cases.

**Comment by davezarzycki** [comment]
I agree with Jordan. StringSwitch is fine in the general case. It just doesn’t scale up when macro meta-programming.

**Comment by davezarzycki** [comment]
PS – I'd be happy to revert this change and commit a fix/workaround to LLVM proper, but (as far as I know) I don't have commit access and nobody has merged my recent reviewed/approved patch to `StringSwitch`. ;-)

For reference, LLVM rebuilds 41% faster on a many-core machine with the change below. Also, I haven't seen any evidence yet that the templated cleverness yields better performance at run time. I think one would need to use genuine macro meta-programming to get the attempted run time performance (which is probably fine for the `#include` and then `#define XYZ(...) .Case(...)` scenarios.

    diff --git i/include/llvm/ADT/StringSwitch.h w/include/llvm/ADT/StringSwitch.h
    index 75577b7738b..328df8fc7d8 100644
    --- i/include/llvm/ADT/StringSwitch.h
    +++ w/include/llvm/ADT/StringSwitch.h
    @@ -69,12 +69,9 @@ public:
       ~StringSwitch() = default;
     
       // Case-sensitive case matchers
    -  template<unsigned N>
       LLVM_ATTRIBUTE_ALWAYS_INLINE
    -  StringSwitch& Case(const char (&S)[N], const T& Value) {
    -    assert(N);
    -    if (!Result && N-1 == Str.size() &&
    -        (N == 1 || std::memcmp(S, Str.data(), N-1) == 0)) {
    +  StringSwitch& Case(const char (S)[], const T& Value) {
    +    if (!Result && S == Str) {
           Result = &Value;
         }
         return *this;

**Comment by davezarzycki** [comment]
If someone could kindly merge the following, then we can revert this change:
https://reviews.llvm.org/D43436
